### ISSUE TITLE
Implement the out of bag prediction in Random Forest mode

### ISSUE THREAD

**Issue by oliviermeslin** [microsoft/LightGBM#6805]
## Summary

LightGBM could support out-of-bag predictions when used in random forest mode (`boosting_type = 'rf'` and `bagging_fraction < 1`). This means that, for a given instance, the prediction of the model would be based only on the trees that were not trained using this particular instance. This feature has been mentioned multiple times (see this [issue](https://github.com/microsoft/LightGBM/issues/1038), this [issue](https://github.com/microsoft/LightGBM/issues/47) and this [issue]( https://github.com/microsoft/LightGBM/issues/5556)).

### What are out-of-bag predictions and why are they useful?
A specific feature of the random forest algorithm (Breiman 2001) is the out-of-bag (OOB) method: with this method it is possible to evaluate the performance of the model _using training data_. The principle is to build for each instance a prediction using only the trees that were trained on a bootstrap subsample that did not contain this particular instance. This method has several practical advantages:

- with OOB predictions, there is no need for a holdout/test set, meaning that all the data can be used to train the model; this is a real gain when working on small datasets;
- OOB predictions can be used as an alternative to cross-validation, implying a large speed-up in training;
- OOB predictions of a random forest can be safely compared to the target on the training data (for instance if you want to detect anomalies by comparing the target with an RF prediction).

OOB predictions are widely used by RF users and are supported in all common implementations of random forests (see for instance [`scikit-learn`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html) and [`ranger`](https://cran.r-project.org/web/packages/ranger/index.html) in `R`.

### Why would out-of-bag predictions be useful in LightGBM?

LightGBM is an excellent implementation of random forests, for two reasons. First, all defining features of random forests have already been implemented in LightGBM (building trees separately instead of sequentially, row subsampling, feature subsampling at the node level), except sampling with replacement but this does not matter much in practice. Second, LightGBM in RF mode is clearly more convenient and much faster than all the common implementations of random forests, because these implementations do not have all the nice features of LightGBM (histogram-based learning, native support of categorical variables, depth-wise tree building, GOSS, custom loss functions...).

Unfortunately, one capital feature is still missing: OOB predictions, and I think it prevents LightGBM from becoming the default way to go for random forests users.

Note: my point is __not about performance__, so the usual argument “`gbdt` is better than `rf`” would not be relevant here. Second, the solution proposed in this [issue](https://github.com/microsoft/LightGBM/issues/5556) (basically using train/validation split) would not answer my concern because it does not offer advantages 1/ and 3/ of OOB predictions.

## Description

The first step towards introducing this feature could consist in keep track of what instances were used to train each tree. I guess this could be done by storing the indices of sampled instances as an attribute of the `booster_` object.

The second step could consist in implementing the OOB prediction itself. An option `oob_prediction =False` could be added to the `predict()` method of LightGBM models (raising an error if `boosting_type` is not `rf`). I’m not familiar with the underlying C++ implementation of prediction in LighGBM, but I understand from older [comments]( https://github.com/microsoft/LightGBM/blob/9f1af051b44564eaab2bebe1612c6a52217bb32b/src/boosting/score_updater.hpp#L95) that some work might have been done in this direction.

**Comment by oliviermeslin** [comment]
@jameslamb : do you have an opinion on this suggestion?

**Comment by jameslamb** [comment]
Thanks very much for the excellent write-up!

<details><summary>side note: I made small edits to your initial post (click me)</summary>

I hope you don't mind, but I've edited it slightly... on GitHub, including bare links to within-the-same-repo content instead of hiding them in markdown hyperlinks like `[this](...)` leads to some nice readability improvements, like rendering commit-anchored code links inline so the code can be viewed without clicking.

</details>

**I am totally convinced that something like this is valuable.**

I'll share some implementation challenges I see, for anyone who is interested in attempting this. This is not meant to discourage development, but just to help shape a plan.

<hr>

## Problem 1 - LightGBM does not keep track of used indices from older iterations

I believe that LightGBM generates a new random subset of `{bagging_fraction} * num_rows` rows each `{bagging_freq}` iterations *(I say "believe" because I am not much of a C++ developer, and even after years of working on LightGBM there are still many parts of its internals I haven't interacted much with)*.

That overwrites the contents of a buffer on the `BaggingSampleStrategy` object *in place*, around here:

https://github.com/microsoft/LightGBM/blob/f91dcfee8baf5fe329bfddadf38577f6443fdf18/src/boosting/bagging.hpp#L163

This means that some modifications to the training code would be required to determine at prediction time which trees were trained on a subset excluding a certain sample.

It'd also introduce a (potentially significant!) new source of memory usage. Maybe some users would be happy to pay that cost in exchange for supporting OOB predictions, but others might not... for backwards-compatibility reasons, we might want to make that tracking configurable based on a parameter, which would increase the implementation complexity further.

<hr>

## Problem 2 - Model files / string representation would require schema changes (and probably would grow significantly larger)

Let's say that during training, the `GBDT` or `TreeLearner` object (or something else) maintained a separate data structure mapping row indices to tree indices over all iteration.

Even if that happened, how would it be supported in model files / strings like those created by `LGBM_BoosterSaveModelToString()`?

LightGBM does not currently have a binary serialization format (https://github.com/microsoft/LightGBM/issues/4681#issuecomment-1008181544), so storing a large number of row indices, even some compressed representation of them, would probably significantly increase model file size.

It might be ok to add initial support with the constraint "OOB predictions do not work with models loaded from text files", but I suspect we'd receive a request to add that support very soon after that initial support was released. Especially given that the string representation is used in other non-file operations, like `copy.deepcopy()` on a `Booster` object in-memory in the Python package:

https://github.com/microsoft/LightGBM/blob/f91dcfee8baf5fe329bfddadf38577f6443fdf18/python-package/lightgbm/basic.py#L3695-L3697

Being able to perform "training continuation" is a really powerful feature of GBDT implementations like LightGBM. I described some different versions of this over here: https://stackoverflow.com/a/73669068/3986677

Maybe that's not too important for OOB predictions, which are mainly for at-training-time evaluation, and maybe it'd be enough to simply not support OOB predictions from a `Booster` object that no longer contained a reference to the training `Dataset`.

But still a consideration that should be accounted for.

<hr>

## Some Other References

Linking some related discussions here:

* https://github.com/catboost/catboost/issues/1443
* https://github.com/dmlc/xgboost/issues/10096#issuecomment-1984009643
* https://github.com/dmlc/xgboost/issues/1197
* https://github.com/dmlc/xgboost/issues/1070
